### PR TITLE
nixos/nginx: fix config validation with hostnames in proxy_pass

### PR DESCRIPTION
**Pull request by symphorien**
###### Description of changes

There are actually many directives which trigger domain name resolution, so using regexes to "fix" them all does not seem like a good approach. Instead I run a fake dns server that answers all requests with a dummy IP.
Unfortunately libredirect tricks to point nginx to a fake /etc/resolv.conf don't work, because nginx uses getaddrinfo, so it is glibc which reads /etc/resolv.conf, and one cannot fake glibc calls to glibc it seems.
So I use bubblewrap instead, but that requires user namespaces, which are not always available. If this is not the case, and validation fails on dns failure, then we ignore the failure. The goal is that if it builds somewhere, it builds everywhere (use case: heterogenous remote builders). I have not yet tested building without user namespaces.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @delroth

### CODE REVIEW

**Comment by symphorien** [comment]
now works with user namespaces disabled

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -1815,6 +1815,8 @@ self: super: with self; {
 
   clize = callPackage ../development/python-modules/clize { };
 
+  cli-formatter = callPackage ../development/python-modules/cli-formatter { };
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix:1818]: Those packages look unrelated.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -2696,6 +2696,8 @@ self: super: with self; {
 
   dnspython = callPackage ../development/python-modules/dnspython { };
 
+  dns-messages = callPackage ../development/python-modules/dns-messages { };
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix:2699]: This one as well.

File: pkgs/tools/networking/fakedns/default.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{ python3, fetchFromGitHub, lib }:
+python3.pkgs.buildPythonApplication {
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/fakedns/default.nix:2]:
```suggestion
{ python3, fetchFromGitHub, lib }:

python3.pkgs.buildPythonApplication {
```

File: pkgs/tools/networking/fakedns/default.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    repo = "fakedns";
+    rev = "11397c3e5180f30de184fba1dbfe9fc8ea85fadd";
+    sha256 = "sha256-ZGNZ8pa9loSTDqrx3ra7sRfsDKzJ1Z2UDlsg+qGq9xo=";
+  };
+
+  propagatedBuildInputs = with python3.pkgs; [
+    cli-formatter
+    dns-messages
+  ];
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/fakedns/default.nix:17]:
```suggestion
  ];

  pythonImportsCheck = [ "fakedns" ];

  meta = {
```

File: pkgs/development/python-modules/dns-messages/default.nix [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  pname = "dns-messages";
+  version = "1.0.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-00gjYwEUeizid/kXxGV2WjBzO/PCpDjV2hMmYUdjPd4=";
+  };
+
+  doCheck = false;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dns-messages/default.nix:12]:
```suggestion

  pythonImportsCheck = [ "dns_messages" ];

```

File: pkgs/development/python-modules/dns-messages/default.nix [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+buildPythonPackage rec {
+  pname = "dns-messages";
+  version = "1.0.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-00gjYwEUeizid/kXxGV2WjBzO/PCpDjV2hMmYUdjPd4=";
+  };
+
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dns-messages/default.nix:11]:
```suggestion
  # has no tests
  doCheck = false;
```

File: pkgs/development/python-modules/dns-messages/default.nix [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{ buildPythonPackage, fetchPypi, lib }:
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/dns-messages/default.nix:2]:
```suggestion
{ buildPythonPackage, fetchPypi, lib }:

buildPythonPackage rec {
```

File: pkgs/development/python-modules/cli-formatter/default.nix [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{ buildPythonPackage, fetchPypi, lib }:
+buildPythonPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cli-formatter/default.nix:2]:
```suggestion
{ buildPythonPackage, fetchPypi, lib }:

buildPythonPackage rec {
```

File: pkgs/development/python-modules/cli-formatter/default.nix [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  pname = "cli-formatter";
+  version = "1.2.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-Vpi2R0DHL9ouxsEtcyJx+reMjItlvtihfbMxJbECfak=";
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cli-formatter/default.nix:10]:
```suggestion

  # has no tests
  doCheck = false;

  pythonImportsCheck = [ "cli_formatter" ];

```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 10]
```diff
@@ -402,24 +402,63 @@ let
...
+    cp conf root/etc/nginx.conf
+    bwrun="bwrap --unshare-net --cap-add CAP_NET_BIND_SERVICE --bind root / --bind /nix /nix"
+    if $bwrun true; then
+      # user namespaces are available
+      userns=yes;
+      $bwrun ${nginxWithResolver} > out 2>&1;
+    else
+      userns=no;
+      echo "user namespaces are not available, dns resolution will fail if required";
+      nginx -t -c $(readlink -f ./conf) > out 2>&1 || true;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:449]:
```suggestion
      userns=yes
      $bwrun ${nginxWithResolver} > out 2>&1
    else
      userns=no
      echo "user namespaces are not available, dns resolution will fail if required"
      nginx -t -c $(readlink -f ./conf) > out 2>&1 || true
```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 11]
```diff
@@ -402,24 +402,63 @@ let
...
+    for i in $(seq 1 1000); do
+      if getent hosts foo.bar; then
+        break
+      fi
+      sleep 1
+    done
+    nginx -t -c /etc/nginx.conf || true
+  '';
+
+  validatedConfigFile = pkgs.runCommand "validated-nginx.conf" { nativeBuildInputs = [ cfg.package pkgs.bubblewrap pkgs.fakedns pkgs.getent ]; } ''
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:426]:
```suggestion
  validatedConfigFile = pkgs.runCommand "validated-nginx.conf" {
    nativeBuildInputs = with pkgs; [ cfg.package bubblewrap fakedns getent ];
  } ''
```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 12]
```diff
@@ -402,24 +402,63 @@ let
...
+    [DomainPattern]
+    name_pattern = *
+    answer_A = 1.2.3.4
+    answer_AAAA = 8::1
+  '';
+  nginxWithResolver = pkgs.writeShellScript "nginxWithResolver" ''
+    set -e
+    fakedns --config ${fakednsConfig} &
+    # wait for the server to be ready
+    for i in $(seq 1 1000); do
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:417]:
```suggestion
    for i in $(seq 1 30); do
```
waiting 1000 seconds for that to come up is to long

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 13]
```diff
@@ -402,24 +402,63 @@ let
...
     openssl x509 -req -days 1 -in server.csr -signkey $out/server.key -out $out/server.crt
   '';
-  validatedConfigFile = pkgs.runCommand "validated-nginx.conf" { nativeBuildInputs = [ cfg.package ]; } ''
+  fakednsConfig = builtins.toFile "fakednsConfig" ''
+    [Settings]
+    ListenOnPort = 53
+    [DomainPattern]
+    name_pattern = *
+    answer_A = 1.2.3.4
+    answer_AAAA = 8::1
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:411]: We should use private addresses or if there are some to be not resolveable them.

**Review by SuperSandro2000 — COMMENTED** [review]
There is not an easier way to run config validation?
We are getting into territoriality where the overhead and workarounds are eventually outweighing the benefits.

**Comment by symphorien** [comment]
I pushed changes which
- solve @Mic92 issue with dhparams
- address some of the stylistic comments of @SuperSandro2000 

>  We are getting into territoriality where the overhead and workarounds are eventually outweighing the benefits.

Yes. I started https://github.com/NixOS/nixpkgs/pull/205561 because nixcloud-webservice has been doing such a validation since quite a long time, but it turns out it is pretty limited for more diverse configs. I can make it work with what is arguably an ugly pile of hacks. If you think this is too hacky, we can revert introducing the validation. I am also hesitating, other opinions welcome.

**Comment by SuperSandro2000** [comment]
> I can make it work with what is arguably an ugly pile of hacks. If you think this is too hacky, we can revert introducing the validation. I am also hesitating, other opinions welcome.

To be honest the cleanest solution would probably be to change the nginx code to better fit our usecase and try to upstream that.

**Comment by symphorien** [comment]
TBH, that's not something I'm willing to attempt.

**Comment by symphorien** [comment]
superseded by https://github.com/NixOS/nixpkgs/pull/209075
